File: src/chapters/numerical-calculus/content/code/composite.cpp
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;

// Composite trapezoidal (T) and Simpson (S) on [a, b] with n subintervals.
void composite(function<double(double)> f, double a, double b, int n, double &T, double &S) {
    if (n % 2) ++n;                          // Simpson needs even n
    double h = (b - a) / n;
    T = (f(a) + f(b)) / 2;
    S = f(a) + f(b);
    for (int i = 1; i < n; ++i) {
        double yi = f(a + i * h);
        T += yi;
        S += (i % 2 ? 4 : 2) * yi;
    }
    T *= h;
    S *= h / 3;
}

int main() {
    cout.precision(7);
    double T, S;
    composite([](double x) { return exp(x); }, 0, 1, 10, T, S);
    cout << fixed << "trapezoid ~ " << T << "\n";
    cout << "Simpson   ~ " << S << "\n";
}
// -> trapezoid ~ 1.7197135, Simpson ~ 1.7182828; exact e-1 = 1.7182818

File: src/chapters/numerical-calculus/content/code/composite.r
# Composite trapezoidal and Simpson rules on [a, b] with n subintervals.
composite <- function(f, a, b, n = 10) {
  if (n %% 2 == 1) n <- n + 1            # Simpson needs even n
  h <- (b - a) / n
  x <- seq(a, b, length.out = n + 1)
  y <- f(x)
  odd <- seq(2, n, by = 2)               # interior odd-position points (1-indexed)
  even <- seq(3, n - 1, by = 2)
  T <- h * (y[1] / 2 + sum(y[2:n]) + y[n + 1] / 2)
  S <- h / 3 * (y[1] + y[n + 1] + 4 * sum(y[odd]) + 2 * sum(y[even]))
  c(trapezoid = T, simpson = S)
}

r <- composite(exp, 0, 1, 10)
cat("trapezoid ~", r["trapezoid"], "\n")
cat("Simpson   ~", r["simpson"], "\n")
# -> trapezoid ~ 1.7197135, Simpson ~ 1.7182828; exact e-1 = 1.7182818
